File: utilslib/src/main/java/com/ashlikun/utils/ui/EditHelper.java
package com.ashlikun.utils.ui;

import android.content.Context;
import android.support.annotation.StringRes;
import android.support.design.widget.TextInputLayout;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.TextView;

import com.ashlikun.utils.animator.AnimUtils;

import java.util.ArrayList;

import static android.R.attr.id;


/**
 * 作者　　: 李坤
 * 创建时间: 2017/6/28 13:30
 * <p>
 * 方法功能：输入框检查工具
 * 1：最后再调用check正则判断
 * 2：可监听输入状态，然后回掉接口判断
 */
public class EditHelper {
    /**
     * 全局默认的是否动画
     */
    public static boolean IS_ANIM = true;
    public Context context;
    private boolean isAnim = IS_ANIM;
    private ArrayList<EditHelperData> mEdithelpdatas;

    public EditHelper(Context context) {
        this.context = context;
    }

    public EditHelper setAnim(boolean anim) {
        isAnim = anim;
        return this;
    }

    /**
     * 作者　　: 李坤
     * 创建时间: 2017/6/28 13:32
     * <p>
     * 方法功能：设置监听的输入框
     *
     * @param edits 多个被检测的EditView对象
     */
    public EditHelper setEditText(EditHelperData... edits) {
        if (mEdithelpdatas == null) {
            mEdithelpdatas = new ArrayList<>();
        }
        mEdithelpdatas.clear();
        for (EditHelperData e : edits) {
            addEditHelperData(e);
        }
        return this;
    }

    /**
     * 作者　　: 李坤
     * 创建时间: 2017/6/28 13:33
     * <p>
     * 方法功能：清空
     */

    public EditHelper clear() {
        if (mEdithelpdatas != null) {
            mEdithelpdatas.clear();
        }
        return this;
    }

    public EditHelper addEditHelperData(EditHelperData edits) {
        if (edits != null) {
            if (mEdithelpdatas == null) {
                mEdithelpdatas = new ArrayList<>();
            }
            mEdithelpdatas.add(edits);
            if (!isAnim) {
                edits.setAnim(false);
            }
            addTextChangedListener(edits);
        }
        return this;
    }

    private void addTextChangedListener(EditHelperData edits) {
        if (edits.getView() instanceof TextInputLayout) {
            TextView v = edits.getTextView();
            if (v != null) {
                v.addTextChangedListener(new EditHelper.MyTextWatcher(edits));
            }
        }
    }


    public EditHelperData getEditHelperData(int index) {
        return mEdithelpdatas.get(index);
    }

    /**
     * 作者　　: 李坤
     * 创建时间: 2017/6/28 13:37
     * <p>
     * 方法功能：检查是否满足
     */
    public boolean check() {
        try {
            if (mEdithelpdatas == null) {
                return false;
            }
            for (int i = 0; i < mEdithelpdatas.size(); i++) {
                EditHelperData e = mEdithelpdatas.get(i);
                if (!e.check(context)) {
                    return false;
                }
            }
            return true;
        } catch (IndexOutOfBoundsException e) {
            e.printStackTrace();
            return true;
        }

    }

    /**
     * 作者　　: 李坤
     * 创建时间: 2017/6/28 13:37
     * <p>
     * 方法功能：检查某个view是否满足
     */

    public boolean check(int index) {
        try {
            if (mEdithelpdatas == null) {
                return false;
            }
            EditHelperData e = mEdithelpdatas.get(id);
            if (e == null || !e.check(context)) {
                return false;
            }
            return true;
        } catch (IndexOutOfBoundsException e) {
            e.printStackTrace();
            return true;
        }

    }

    public static class EditHelperData {

        View view;
        String msg;
        String regex;
        boolean isAnim = true;

        public EditHelperData(View textView, String regex, String msg) {
            this.regex = regex;
            this.view = textView;
            this.msg = msg;
        }

        public EditHelperData(View textView, String regex, @StringRes int msgStringId) {
            this.regex = regex;
            this.view = textView;
            this.msg = textView.getContext().getString(msgStringId);
        }


        public EditHelperData(View textView, int regexMaxLenght, @StringRes int msgStringId) {
            this.regex = "[\\S]{1," + regexMaxLenght + "}";
            this.view = textView;
            this.msg = textView.getContext().getString(msgStringId);
        }

        public EditHelperData setAnim(boolean anim) {
            isAnim = anim;
            return this;
        }

        public TextView getTextView() {

            if (view instanceof TextInputLayout) {
                return ((TextInputLayout) view).getEditText();
            } else if (view instanceof TextView) {
                return (TextView) view;
            } else {
                return null;
            }
        }

        public View getView() {
            return view;
        }

        public EditHelperData setView(View view) {
            this.view = view;
            return this;
        }

        public String getMsg() {
            return msg;
        }

        public EditHelperData setMsg(String msg) {
            this.msg = msg;
            return this;
        }

        public EditHelperData setMsg(Context context, @StringRes int msgStringId) {
            this.msg = context.getString(msgStringId);
            return this;
        }

        public String getRegex() {
            return regex;
        }

        public EditHelperData setRegex(String regex) {
            this.regex = regex;
            return this;
        }


        public boolean check(Context context) {
            if (view == null || !getTextView().getText().toString().matches(regex)) {
                if (view instanceof TextInputLayout && ((TextInputLayout) view).isErrorEnabled()) {
                    ((TextInputLayout) view).setError(msg);
                    if (isAnim) {
                        AnimUtils.shakeLeft(view, 0.95f, 3f);
                    }
                } else {
                    SuperToast.get(msg).warn();
                    if (isAnim) {
                        AnimUtils.shakeLeft(view, 0.85f, 6f);
                    }
                }
                view.requestFocus();

                return false;
            }
            if (view instanceof TextInputLayout && ((TextInputLayout) view).isErrorEnabled()) {
                ((TextInputLayout) view).setError("");
            }
            return true;
        }

    }

    public class MyTextWatcher implements TextWatcher {

        EditHelperData editHelperData;

        public MyTextWatcher(EditHelperData editHelperData) {
            this.editHelperData = editHelperData;
        }

        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {

        }

        @Override
        public void afterTextChanged(Editable s) {
            if (editHelperData.getView() instanceof TextInputLayout) {
                int maxLength = ((TextInputLayout) editHelperData.getView()).getCounterMaxLength();
                if (maxLength < s.length()) {
                    editHelperData.check(EditHelper.this.context);
                } else {
                    ((TextInputLayout) editHelperData.getView()).setError("");
                }
            }
        }
    }
}

File: utilslib/src/main/java/com/ashlikun/utils/ui/EditHelper.kt
package com.ashlikun.utils.ui

import android.content.Context
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.TextView
import androidx.annotation.StringRes
import com.ashlikun.utils.animator.AnimUtils.startShakeLeft
import com.ashlikun.utils.ui.modal.SuperToast
import com.google.android.material.textfield.TextInputLayout
import java.util.*

/**
 * 作者　　: 李坤
 * 创建时间: 2017/6/28 13:30
 *
 *
 * 输入框检查工具
 * 1：最后再调用check正则判断
 * 2：可监听输入状态，然后回掉接口判断
 */
class EditHelper(var context: Context) {
    companion object {
        /**
         * 全局默认的是否动画
         */
        var IS_ANIM = true
    }

    var isAnim = IS_ANIM

    //显示toast回调
    var showToas: ((msg: String) -> Unit)? = null

    //显示的toast类型
    var toastType = SuperToast.Info
    private var mEdithelpdatas: ArrayList<EditHelperData>? = null


    /**
     * 设置监听的输入框
     *
     * @param edits 多个被检测的EditView对象
     */
    fun setEditText(vararg edits: EditHelperData?): EditHelper {
        if (mEdithelpdatas == null) {
            mEdithelpdatas = ArrayList()
        }
        mEdithelpdatas!!.clear()
        for (e in edits) {
            addEditHelperData(e)
        }
        return this
    }

    /**
     * 清空
     */
    fun clear(): EditHelper {
        if (mEdithelpdatas != null) {
            mEdithelpdatas!!.clear()
        }
        return this
    }

    fun addEditHelperData(edits: EditHelperData?): EditHelper {
        if (edits != null) {
            if (mEdithelpdatas == null) {
                mEdithelpdatas = ArrayList()
            }
            mEdithelpdatas!!.add(edits)
            if (!isAnim) {
                edits.setAnim(false)
            }

            edits.helper = this
            addTextChangedListener(edits)
        }
        return this
    }

    private fun addTextChangedListener(edits: EditHelperData) {
        if (edits.view is TextInputLayout) {
            val v = edits.textView
            v?.addTextChangedListener(MyTextWatcher(edits))
        }
    }

    fun getEditHelperData(index: Int): EditHelperData {
        return mEdithelpdatas!![index]
    }

    /**
     * 检查是否满足
     */
    fun check(): Boolean {
        return try {
            if (mEdithelpdatas == null) {
                return false
            }
            for (i in mEdithelpdatas!!.indices) {
                val e = mEdithelpdatas!![i]
                if (!e.check(context)) {
                    return false
                }
            }
            true
        } catch (e: IndexOutOfBoundsException) {
            e.printStackTrace()
            true
        }
    }

    /**
     * 检查某个view是否满足
     */
    fun check(index: Int): Boolean {
        return try {
            if (mEdithelpdatas == null) {
                return false
            }
            val e = mEdithelpdatas!![index]
            !(e == null || !e.check(context))
        } catch (e: IndexOutOfBoundsException) {
            e.printStackTrace()
            true
        }
    }

    /**
     * 检查某个view是否满足
     */
    fun check(editView: View): Boolean {
        return try {
            if (mEdithelpdatas == null) {
                return false
            }
            val e = mEdithelpdatas!!.find { it.view == editView }
            !(e == null || !e.check(context))
        } catch (e: IndexOutOfBoundsException) {
            e.printStackTrace()
            true
        }
    }

    class EditHelperData {
        var view: View
        var msg: String
        var regex: String
        var isAnim = true
        lateinit var helper: EditHelper

        constructor(textView: View, regex: String, msg: String) {
            this.regex = regex
            view = textView
            this.msg = msg
        }

        constructor(textView: View, regex: String, @StringRes msgStringId: Int) {
            this.regex = regex
            view = textView
            msg = textView.context.getString(msgStringId)
        }

        constructor(textView: View, regexMaxLenght: Int, @StringRes msgStringId: Int) {
            regex = "[\\S]{1,$regexMaxLenght}"
            view = textView
            msg = textView.context.getString(msgStringId)
        }

        fun setAnim(anim: Boolean): EditHelperData {
            isAnim = anim
            return this
        }

        val textView: TextView?
            get() = if (view is TextInputLayout) {
                (view as TextInputLayout).editText
            } else if (view is TextView) {
                view as TextView?
            } else {
                null
            }

        fun setView(view: View): EditHelperData {
            this.view = view
            return this
        }

        fun setMsg(msg: String): EditHelperData {
            this.msg = msg
            return this
        }

        fun setMsg(context: Context, @StringRes msgStringId: Int): EditHelperData {
            msg = context.getString(msgStringId)
            return this
        }

        fun setRegex(regex: String): EditHelperData {
            this.regex = regex
            return this
        }

        fun check(context: Context): Boolean {
            if (view == null || !textView!!.text.toString().matches(regex.toRegex())) {
                if (view is TextInputLayout && (view as TextInputLayout).isErrorEnabled) {
                    (view as TextInputLayout).error = msg
                    if (isAnim) {
                        startShakeLeft(view, 0.95f, 3f)
                    }
                } else {
                    if (helper.showToas != null) {
                        helper.showToas!!.invoke(msg)
                    } else {
                        SuperToast[msg].setType(helper.toastType).show()
                    }
                    if (isAnim) {
                        startShakeLeft(view!!, 0.85f, 6f)
                    }
                }
                view!!.requestFocus()
                return false
            }
            if (view is TextInputLayout && (view as TextInputLayout).isErrorEnabled) {
                (view as TextInputLayout).error = ""
            }
            return true
        }
    }

    inner class MyTextWatcher(var editHelperData: EditHelperData) : TextWatcher {
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
        override fun afterTextChanged(s: Editable) {
            if (editHelperData.view is TextInputLayout) {
                val maxLength = (editHelperData.view as TextInputLayout?)!!.counterMaxLength
                if (maxLength < s.length) {
                    editHelperData.check(context)
                } else {
                    (editHelperData.view as TextInputLayout?)!!.error = ""
                }
            }
        }
    }


}